## 📚 Étape 0 : Import de bibliothéque
---

In [119]:
import pandas as pd
import os
import numpy as np
import matplotlib as plt
from sklearn.linear_model import LinearRegression

## 🔍 Étape 1 : Analyse du fichier lille_2022.csv
---

In [120]:
import os
print(os.getcwd())

/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/notebooks


In [121]:
df = pd.read_csv("../data/bordeaux_2022.csv")
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13200 entries, 0 to 13199
Data columns (total 44 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Identifiant de document     0 non-null      float64
 1   Reference document          0 non-null      float64
 2   1 Articles CGI              0 non-null      float64
 3   2 Articles CGI              0 non-null      float64
 4   3 Articles CGI              0 non-null      float64
 5   4 Articles CGI              0 non-null      float64
 6   5 Articles CGI              0 non-null      float64
 7   No disposition              13200 non-null  int64  
 8   Date mutation               13200 non-null  object 
 9   Nature mutation             13200 non-null  object 
 10  Valeur fonciere             13200 non-null  float64
 11  No voie                     13155 non-null  float64
 12  B/T/Q                       402 non-null    object 
 13  Type de voie                131

Identifiant de document  Reference document  1 Articles CGI  \
0                      NaN                 NaN             NaN   
1                      NaN                 NaN             NaN   
2                      NaN                 NaN             NaN   
3                      NaN                 NaN             NaN   
4                      NaN                 NaN             NaN   

   2 Articles CGI  3 Articles CGI  4 Articles CGI  5 Articles CGI  \
0             NaN             NaN             NaN             NaN   
1             NaN             NaN             NaN             NaN   
2             NaN             NaN             NaN             NaN   
3             NaN             NaN             NaN             NaN   
4             NaN             NaN             NaN             NaN   

   No disposition Date mutation Nature mutation  ...  Nombre de lots  \
0               1    04/01/2022           Vente  ...               0   
1               1    05/01/2022           Vente  ...               1   
2               1    07/01/2022           Vente  ...               1   
3               1    07/01/2022           Vente  ...               1   
4               1    07/01/2022           Vente  ...               1   

   Code type local   Type local Identifiant local Surface reelle bati  \
0              1.0       Maison               NaN                71.0   
1              2.0  Appartement               NaN                40.0   
2              3.0   Dépendance               NaN                 0.0   
3              2.0  Appartement               NaN                35.0   
4              3.0   Dépendance               NaN                 0.0   

  Nombre pieces principales  Nature culture Nature culture speciale  \
0                       4.0               S                     NaN   
1                       1.0             NaN                     NaN   
2                       0.0             NaN                     NaN   
3                       1.0             NaN                     NaN   
4                       0.0             NaN                     NaN   

   Surface terrain      prix_m2  
0             70.0  5577.464789  
1              NaN  3962.500000  
2              NaN          inf  
3              NaN  5342.857143  
4              NaN          inf  

[5 rows x 44 columns]

## 🧬 Étape 2 : Filtrer les biens de 4 pièces uniquement
---

In [122]:
# Vérification  des colonnes du dataframe pour être sûr du nom exact
print(df.columns)

# Filtrage sur 4.0 (float)
df_4_pieces = df[df['Nombre pieces principales'] == 4].copy()

# Vérifier que le filtrage a fonctionné
print(df_4_pieces.shape)
print(df_4_pieces.head())


Index(['Identifiant de document', 'Reference document', '1 Articles CGI',
       '2 Articles CGI', '3 Articles CGI', '4 Articles CGI', '5 Articles CGI',
       'No disposition', 'Date mutation', 'Nature mutation', 'Valeur fonciere',
       'No voie', 'B/T/Q', 'Type de voie', 'Code voie', 'Voie', 'Code postal',
       'Commune', 'Code departement', 'Code commune', 'Prefixe de section',
       'Section', 'No plan', 'No Volume', '1er lot',
       'Surface Carrez du 1er lot', '2eme lot', 'Surface Carrez du 2eme lot',
       '3eme lot', 'Surface Carrez du 3eme lot', '4eme lot',
       'Surface Carrez du 4eme lot', '5eme lot', 'Surface Carrez du 5eme lot',
       'Nombre de lots', 'Code type local', 'Type local', 'Identifiant local',
       'Surface reelle bati', 'Nombre pieces principales', 'Nature culture',
       'Nature culture speciale', 'Surface terrain', 'prix_m2'],
      dtype='object')
(1071, 44)
    Identifiant de document  Reference document  1 Articles CGI  \
0                   

In [123]:
print(df['Nombre pieces principales'].dtype)
print(df['Nombre pieces principales'].unique())

float64
[ 4.  1.  0.  3.  7.  2.  5.  6.  8. 10.  9. 11. 12. 15. 13.]


## 🧪 Étape 3 : Création de jeu de données distincs pour les appartements et les maisons 
---

### ✅ Jeu de données pour les appartements :



In [124]:
# # Filtrer le DataFrame pour ne garder que les appartements
df_appartements = df[df['Type local'] == 'Appartement'].copy()

# Afficher la liste des colonnes pour vérifier les colonnes disponibles
print(df_appartements.columns)

# Sélection dela liste des colonnes que l'on souhaite conserver
colonnes_a_garder = [
    'Surface reelle bati',
    'Nombre pieces principales',
    'Type local',
    'Surface terrain',     
    'Nombre de lots',
    'prix_m2'
]

# Filtrer cette liste pour ne garder que les colonnes qui existent vraiment dans le DataFrame
colonnes_disponibles = [col for col in colonnes_a_garder if col in df_appartements.columns]

# Sélectionner dans le DataFrame uniquement ces colonnes disponibles
df_appartements = df_appartements[colonnes_disponibles].copy()

# Afficher un aperçu pour vérifier
print(df_appartements.shape)
print(df_appartements.head())



Index(['Identifiant de document', 'Reference document', '1 Articles CGI',
       '2 Articles CGI', '3 Articles CGI', '4 Articles CGI', '5 Articles CGI',
       'No disposition', 'Date mutation', 'Nature mutation', 'Valeur fonciere',
       'No voie', 'B/T/Q', 'Type de voie', 'Code voie', 'Voie', 'Code postal',
       'Commune', 'Code departement', 'Code commune', 'Prefixe de section',
       'Section', 'No plan', 'No Volume', '1er lot',
       'Surface Carrez du 1er lot', '2eme lot', 'Surface Carrez du 2eme lot',
       '3eme lot', 'Surface Carrez du 3eme lot', '4eme lot',
       'Surface Carrez du 4eme lot', '5eme lot', 'Surface Carrez du 5eme lot',
       'Nombre de lots', 'Code type local', 'Type local', 'Identifiant local',
       'Surface reelle bati', 'Nombre pieces principales', 'Nature culture',
       'Nature culture speciale', 'Surface terrain', 'prix_m2'],
      dtype='object')
(5353, 6)
    Surface reelle bati  Nombre pieces principales   Type local  \
1                  40

### ✅ Jeu de données pour les maisons : 



In [125]:
# # Filtrer le DataFrame pour ne garder que les maisons
df_maisons = df[df['Type local'] == 'Maison'].copy()

# Afficher la liste des colonnes pour vérifier les colonnes disponibles
print(df_maisons.columns)

# Sélection dela liste des colonnes que l'on souhaite conserver
colonnes_a_garder = [
    'Surface reelle bati',
    'Nombre pieces principales',
    'Type local',
    'Surface terrain',     
    'Nombre de lots',
    'prix_m2'
]

# Filtrer cette liste pour ne garder que les colonnes qui existent vraiment dans le DataFrame
colonnes_disponibles = [col for col in colonnes_a_garder if col in df_maisons.columns]

# Sélectionner dans le DataFrame uniquement ces colonnes disponibles
df_maisons = df_maisons[colonnes_disponibles].copy()

# Afficher un aperçu pour vérifier
print(df_maisons.head())



Index(['Identifiant de document', 'Reference document', '1 Articles CGI',
       '2 Articles CGI', '3 Articles CGI', '4 Articles CGI', '5 Articles CGI',
       'No disposition', 'Date mutation', 'Nature mutation', 'Valeur fonciere',
       'No voie', 'B/T/Q', 'Type de voie', 'Code voie', 'Voie', 'Code postal',
       'Commune', 'Code departement', 'Code commune', 'Prefixe de section',
       'Section', 'No plan', 'No Volume', '1er lot',
       'Surface Carrez du 1er lot', '2eme lot', 'Surface Carrez du 2eme lot',
       '3eme lot', 'Surface Carrez du 3eme lot', '4eme lot',
       'Surface Carrez du 4eme lot', '5eme lot', 'Surface Carrez du 5eme lot',
       'Nombre de lots', 'Code type local', 'Type local', 'Identifiant local',
       'Surface reelle bati', 'Nombre pieces principales', 'Nature culture',
       'Nature culture speciale', 'Surface terrain', 'prix_m2'],
      dtype='object')
    Surface reelle bati  Nombre pieces principales Type local  \
0                  71.0          

## 🧹 Étape 5 : Nettoyage des données 
---

### ✅ Supprimer les lignes avec valeurs manquantes



In [126]:
from sklearn.preprocessing import LabelEncoder

# 🧼 Étape 1 : Colonnes utiles pour l'entraînement
colonnes_utiles = [
    'Surface reelle bati',
    'Nombre pieces principales',
    'Type local',
    'Surface terrain',
    'Nombre de lots',
    'prix_m2'
]

# 🧹 Étape 2 : Nettoyage des données avec dropna, et création d'une copie propre
df_appartements_clean = df_appartements.dropna(subset=colonnes_utiles).copy()
df_maisons_clean = df_maisons.dropna(subset=colonnes_utiles).copy() 

# 🏷️ Étape 3 : Encodage de 'Type local' sur chaque DataFrame (si besoin de garder les deux)
label_encoder = LabelEncoder()

# Mapping d'encodage global
mapping_type_local = {
    "Appartement": 0,
    "Maison": 1
}

df_appartements_clean["Type local encoded"] = df_appartements_clean["Type local"].map(mapping_type_local)
df_maisons_clean["Type local encoded"] = df_maisons_clean["Type local"].map(mapping_type_local)

# 🧺 Étape 4 : Optionnel – supprimer la colonne originale
df_appartements_clean.drop(columns=["Type local"], inplace=True)
df_maisons_clean.drop(columns=["Type local"], inplace=True)


### ✅ Retirer les valeurs aberrantes (prix_m2 trop faible ou trop élevé)

### 🏢 Outliers appartements 

In [127]:
def detect_outliers(df, column='prix_m2'):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return Q1, Q3, lower_bound, upper_bound, outliers

# Détection des outliers dans les appartements
print("🔎 Détection des outliers sur la colonne 'prix_m2' pour les appartements 🏢")

Q1_appart, Q3_appart, lower_appart, upper_appart, outliers_appart = detect_outliers(df_appartements_clean)

print(f"Q1 : {Q1_appart}, Q3 : {Q3_appart}")
print(f"Borne inférieure : {lower_appart}, Borne supérieure : {upper_appart}")
print(f"Nombre d'outliers détectés : {outliers_appart.shape[0]}\n")

# Dimensions avant nettoyage
print(f"Dimensions avant nettoyage : {df_appartements_clean.shape}")

# Création d'une copie propre sans les outliers
df_appartements_cleaned = df_appartements_clean[
    (df_appartements_clean['prix_m2'] >= lower_appart) &
    (df_appartements_clean['prix_m2'] <= upper_appart)
].copy()

# Dimensions après nettoyage
print(f"Dimensions après nettoyage : {df_appartements_clean.shape}")


🔎 Détection des outliers sur la colonne 'prix_m2' pour les appartements 🏢
Q1 : 11259.328358208955, Q3 : 30232.558139534885
Borne inférieure : -17200.51631377994, Borne supérieure : 58692.402811523774
Nombre d'outliers détectés : 58

Dimensions avant nettoyage : (971, 6)
Dimensions après nettoyage : (971, 6)


In [128]:
df_appartements_clean.head ()

Surface reelle bati  Nombre pieces principales  Surface terrain  \
24                 36.0                        2.0             63.0   
25                 36.0                        2.0             63.0   
26                 26.0                        1.0             63.0   
86                 39.0                        2.0             57.0   
87                 40.0                        2.0             57.0   

    Nombre de lots       prix_m2  Type local encoded  
24               0  23627.777778                   0  
25               0  23627.777778                   0  
26               0  32715.384615                   0  
86               0  15897.435897                   0  
87               0  15500.000000                   0

### 🏡 Outliers maisons 

In [129]:
def detect_outliers(df, column='prix_m2'):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]
    return Q1, Q3, lower_bound, upper_bound, outliers

# Détection des outliers pour les maisons
print("🔍 Détection des outliers sur la colonne 'prix_m2' pour les maisons 🏡")

Q1_maisons, Q3_maisons, lower_maisons, upper_maisons, outliers_maisons = detect_outliers(df_maisons_clean)

print(f"Q1 : {Q1_maisons}, Q3 : {Q3_maisons}")
print(f"Borne inférieure : {lower_maisons}, Borne supérieure : {upper_maisons}")
print(f"Nombre d'outliers détectés : {outliers_maisons.shape[0]}\n")

# Dimensions avant nettoyage
print(f"Dimensions avant nettoyage : {df_maisons_clean.shape}")

# Création du DataFrame nettoyé
df_maisons_cleaned = df_maisons_clean[
    (df_maisons_clean['prix_m2'] >= lower_maisons) &
    (df_maisons_clean['prix_m2'] <= upper_maisons)
].copy()

# Dimensions après nettoyage
print(f"Dimensions après nettoyage : {df_maisons_clean.shape}")


🔍 Détection des outliers sur la colonne 'prix_m2' pour les maisons 🏡
Q1 : 4608.032728141424, Q3 : 6610.712351029252
Borne inférieure : 1604.0132938096813, Borne supérieure : 9614.731785360995
Nombre d'outliers détectés : 104

Dimensions avant nettoyage : (1396, 6)
Dimensions après nettoyage : (1396, 6)


In [130]:
df_maisons_cleaned.head()

Surface reelle bati  Nombre pieces principales  Surface terrain  \
0                  71.0                        4.0             70.0   
5                  65.0                        3.0             45.0   
10                162.0                        7.0            111.0   
31                 92.0                        4.0            121.0   
44                 82.0                        3.0             44.0   

    Nombre de lots      prix_m2  Type local encoded  
0                0  5577.464789                   1  
5                0  4807.692308                   1  
10               0  3117.283951                   1  
31               0  5760.869565                   1  
44               0  4939.024390                   1

## 🤖 Étape 7 : Préparation des données pour l'entrainement
---

### ✅ Préparation des données pour l'entrainement des appartements 

In [131]:

# Variables explicatives (X) : toutes les colonnes sauf la cible
X_appart = df_appartements_cleaned.drop(columns=['prix_m2'])

# Variable cible (y)
y_appart = df_appartements_cleaned['prix_m2']


# Vérification des jeux d'entraînement et de test
print("Taille X :", X_appart.shape)
print("Taille y :", y_appart.shape)


Taille X : (913, 5)
Taille y : (913,)


### ✅ Préparation des données pour l'entrainement des maisons

In [132]:
from sklearn.model_selection import train_test_split

# Variables explicatives (X) : toutes les colonnes sauf la cible
X_maisons = df_maisons_cleaned.drop(columns=['prix_m2'])

# Variable cible (y)
y_maisons = df_maisons_cleaned['prix_m2']



# Vérification des jeux d'entraînement et de test
print("Taille de X :", X_maisons.shape)
print("Taille de y :", y_maisons.shape)


Taille de X : (1292, 5)
Taille de y : (1292,)


### ⚙️ Standardisation des données avec StandardScaler 
---


### 🏢 StandardScaler pour les appartements : 

In [133]:
from sklearn.preprocessing import StandardScaler

# Initialiser le scaler
scaler = StandardScaler()

# Fit sur l'entraînement + transformation
X_appart_scaled = scaler.fit_transform(X_appart)

print(X_appart.shape)
print(X_appart_scaled.shape)

(913, 5)
(913, 5)


### 🏡 StandardScaler  pour les maisons : 

In [134]:
from sklearn.preprocessing import StandardScaler

# Initialiser le scaler
scaler = StandardScaler()

# Transformation du test
X_maisons_scaled = scaler.fit_transform(X_maisons)

print(X_maisons.shape)
print(X_maisons_scaled.shape)

(1292, 5)
(1292, 5)


## 🤖 Étape 8 : Application des modéles entrainé sur Lille
---

In [135]:
import os

# Vérification du répertoire courant et du chemin absolu du dossier models
print("📌 Répertoire courant :", os.getcwd())
print("📁 Répertoire absolu du dossier models :", os.path.abspath("models/"))


📌 Répertoire courant : /Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/notebooks
📁 Répertoire absolu du dossier models : /Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/notebooks/models


### 🏢 🔄 Chargement des modéles pour les appartements : 

In [140]:
import joblib

# Rechargement des modèles (appartements uniquement)

LinearRegression_appart = joblib.load("../models/linear_regression_apparts_Lille.pkl")
DecisionTree_appart = joblib.load("../models/decision_tree_apparts_Lille.pkl")
DecisionTree_GridSearchCV_appart = joblib.load("../models/decision_tree_gridsearchcv_apparts_Lille.pkl")
RandomForest_appart = joblib.load("../models/random_forest_apparts_Lille.pkl")
RandomForest_GridSearchCV_appart = joblib.load("../models/random_forest_gridsearchcv_apparts_Lille.pkl")
XGBR_appart = joblib.load("../models/xgboost_apparts_Lille.pkl")



### 🏡 🔄 Chargement des modéles pour les maisons : 

In [139]:
import joblib

# Rechargement des modèles (maisons uniquement)

DecisionTree_maisons = joblib.load("../models/decision_tree_maisons_Lille.pkl")
DecisionTree_GridSearchCV_maisons = joblib.load("../models/decision_tree_gridsearchcv_maisons_Lille.pkl")
LinearRegression_maisons = joblib.load("../models/linear_regression_maisons_Lille.pkl")
RandomForest_maisons = joblib.load("../models/random_forest_maisons_Lille.pkl")
RandomForest_GridSearchCV_maisons = joblib.load("../models/random_forest_gridsearchcv_maisons_Lille.pkl")
XGBR_maisons = joblib.load("../models/xgboost_maisons_Lille.pkl")

## 📂🔄 Étape 9 : Prédictions sur les données de Bordeaux 
---

### 🎯 🏢 Prédictions sur Bordeaux pour les appartement 

In [141]:
# Appartements Bordeaux

y_pred_lr_appart = LinearRegression_appart.predict(X_appart_scaled)
y_pred_dtr_appart = DecisionTree_appart.predict(X_appart_scaled)
y_pred_dtr_gsr_appart = DecisionTree_GridSearchCV_appart.predict(X_appart_scaled)
y_pred_rf_appart = RandomForest_appart.predict(X_appart_scaled)
y_pred_rf_gsr_appart = RandomForest_GridSearchCV_appart.predict(X_appart_scaled)
y_pred_xgbr_appart = XGBR_appart.predict(X_appart_scaled)



/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(
/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(
/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(
/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.v

In [142]:
#  Vérifier que ça matche
print("🎯 y_reel :", y_appart.shape)
print("🧠 y_pred :", y_pred_lr_appart.shape)
print("🧠 y_pred :", y_pred_dtr_appart.shape)
print("🧠 y_pred :", y_pred_dtr_gsr_appart.shape)
print("🧠 y_pred :", y_pred_rf_appart.shape)
print("🧠 y_pred :", y_pred_rf_gsr_appart.shape)
print("🧠 y_pred :", y_pred_xgbr_appart.shape)

🎯 y_reel : (913,)
🧠 y_pred : (913,)
🧠 y_pred : (913,)
🧠 y_pred : (913,)
🧠 y_pred : (913,)
🧠 y_pred : (913,)
🧠 y_pred : (913,)


### 🎯 🏡 Prédictions sur Bordeaux pour les maisons 

In [143]:
# Maisons Bordeaux

y_pred_lr_maisons = LinearRegression_maisons.predict(X_maisons_scaled)
y_pred_dtr_maisons = DecisionTree_maisons.predict(X_maisons_scaled)
y_pred_dtr_gsr_maisons = DecisionTree_GridSearchCV_maisons.predict(X_maisons_scaled)
y_pred_rf_maisons = RandomForest_maisons.predict(X_maisons_scaled)
y_pred_rf_gsr_maisons = RandomForest_GridSearchCV_maisons.predict(X_maisons_scaled)
y_pred_xgbr_maisons = XGBR_maisons.predict(X_maisons_scaled)

/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(
/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(
/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
  warnings.warn(
/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.venv/lib/python3.12/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(
/Users/aminaabdm/Desktop/FastAPI_IA_Immobilier/.v

In [144]:
#  Vérifier que ça matche
print("🎯 y_reel :", y_maisons.shape)
print("🧠 y_pred :", y_pred_lr_maisons.shape)
print("🧠 y_pred :", y_pred_dtr_maisons.shape)
print("🧠 y_pred :", y_pred_dtr_gsr_maisons.shape)
print("🧠 y_pred :", y_pred_rf_maisons.shape)
print("🧠 y_pred :", y_pred_rf_gsr_maisons.shape)
print("🧠 y_pred :", y_pred_xgbr_maisons.shape)

🎯 y_reel : (1292,)
🧠 y_pred : (1292,)
🧠 y_pred : (1292,)
🧠 y_pred : (1292,)
🧠 y_pred : (1292,)
🧠 y_pred : (1292,)
🧠 y_pred : (1292,)


### 📋 🏢 Afficher les prédictions de chaque modéle pour les appartements 

In [145]:
from sklearn.metrics import r2_score, mean_squared_error
import pandas as pd

# 🧮 Calculs Appartements
r2_lr_appart  = r2_score(y_appart, y_pred_lr_appart)
r2_dtr_appart = r2_score(y_appart, y_pred_dtr_appart)
r2_dtr_gsr_appart = r2_score(y_appart, y_pred_dtr_gsr_appart)
r2_rfr_appart = r2_score(y_appart, y_pred_rf_appart)
r2_gsr_rfr_appart = r2_score(y_appart, y_pred_rf_gsr_appart)
r2_xgbr_appart = r2_score(y_appart, y_pred_xgbr_appart)

mse_lr_appart  = mean_squared_error(y_appart, y_pred_lr_appart)
mse_dtr_appart = mean_squared_error(y_appart, y_pred_dtr_appart)
mse_gsr_dtr_appart = mean_squared_error(y_appart, y_pred_dtr_gsr_appart)
mse_rfr_appart = mean_squared_error(y_appart, y_pred_rf_appart)
mse_gsr_rfr_appart = mean_squared_error(y_appart, y_pred_rf_gsr_appart)
mse_xgbr_appart = mean_squared_error(y_appart, y_pred_xgbr_appart)



# 📊 Résultats pour les appartements
resultats_appart = {
    "LinearRegression": {"r2": r2_lr_appart, "mse": mse_lr_appart},
    "DecisionTree": {"r2": r2_dtr_appart, "mse": mse_dtr_appart},
    "DecisionTree (GridSearchCV)" : {"r2": r2_dtr_gsr_appart, "mse": mse_gsr_dtr_appart},
    "RandomForest": {"r2": r2_rfr_appart, "mse": mse_rfr_appart},
    "RandomForest (GridSearchCV)": {"r2": r2_gsr_rfr_appart, "mse": mse_gsr_rfr_appart},
    "XGBoost": {"r2": r2_xgbr_appart, "mse": mse_xgbr_appart}
}



# 🧾 Convertion en DataFrames
df_resultats_appart = pd.DataFrame(resultats_appart).T


# 📣 Affichage
print("\n🏢 Résultats pour les appartements :")
display(df_resultats_appart.sort_values(by='r2', ascending=False))




🏢 Résultats pour les appartements :


r2           mse
LinearRegression             0.005896  1.732843e+08
RandomForest (GridSearchCV) -0.020400  1.778682e+08
RandomForest                -0.028668  1.793094e+08
XGBoost                     -0.168176  2.036273e+08
DecisionTree                -0.327225  2.313514e+08
DecisionTree (GridSearchCV) -0.374003  2.395055e+08

In [146]:
# 🧮 Calculs Maisons


r2_lr_maisons  = r2_score(y_maisons, y_pred_lr_maisons)
r2_dtr_maisons = r2_score(y_maisons, y_pred_dtr_maisons)
r2_dtr_gsr_maisons = r2_score(y_maisons, y_pred_dtr_gsr_maisons)
r2_rfr_maisons = r2_score(y_maisons, y_pred_rf_maisons)
r2_gsr_rfr_maisons = r2_score(y_maisons, y_pred_rf_gsr_maisons)
r2_xgbr_maison  = r2_score(y_maisons, y_pred_xgbr_maisons)

mse_lr_maison  = mean_squared_error(y_maisons, y_pred_lr_maisons)
mse_dtr_maisons = mean_squared_error(y_maisons, y_pred_dtr_maisons)
mse_gsr_dtr_maisons = mean_squared_error(y_maisons, y_pred_dtr_gsr_maisons)
mse_rfr_maisons = mean_squared_error(y_maisons, y_pred_rf_maisons)
mse_gsc_rfr_maisons = mean_squared_error(y_maisons, y_pred_rf_gsr_maisons)
mse_xgbr_maison = mean_squared_error(y_maisons, y_pred_xgbr_maisons)

# 📊 Résultats pour les maisons
resultats_maisons = {
    "LinearRegression": {"r2": r2_lr_maisons, "mse": mse_lr_maison},
    "DecisionTree": {"r2": r2_dtr_maisons, "mse": mse_dtr_maisons},
    "DecisionTree (GridSearchCV)" :{"r2": r2_dtr_gsr_appart, "mse": mse_gsr_dtr_maisons},
    "RandomForest": {"r2": r2_rfr_maisons, "mse": mse_rfr_maisons},
    "RandomForest (GridSearchCV)": {"r2": r2_gsr_rfr_maisons, "mse": mse_gsc_rfr_maisons},
    "XGBoost": {"r2": r2_xgbr_maison, "mse": mse_xgbr_maison}
}

# 🧾 Convertion en DataFrames
df_resultats_maisons = pd.DataFrame(resultats_maisons).T

# 📣 Affichage
print("\n🏠 Résultats pour les maisons :")
display(df_resultats_maisons.sort_values(by='r2', ascending=False))



🏠 Résultats pour les maisons :


r2           mse
DecisionTree (GridSearchCV) -0.374003  8.130029e+06
RandomForest (GridSearchCV) -2.956385  8.187240e+06
RandomForest                -2.973259  8.222158e+06
LinearRegression            -3.040591  8.361494e+06
XGBoost                     -3.163592  8.616030e+06
DecisionTree                -3.517995  9.349422e+06

## 📊  Étape 10 : Comparaisons des performances entre Lille et Bordeaux  
---

### 🧪 Étude comparative des modèles de prédiction immobilière : Lille vs Bordeaux

### 🏢 Résultats des modèles Appartements 


| Modèle                         | R² Lille | MSE Lille       | R² Bordeaux | MSE Bordeaux     |
|-------------------------------|----------|------------------|-------------|------------------|
| XGBoost                       | 0.680    | 32 993 290       | -0.168      | 203 627 300      |
| RandomForest                  | 0.647    | 36 441 400       | -0.029      | 179 309 400      |
| RandomForest (GridSearchCV)  | 0.645    | 36 555 040       | -0.020      | 177 868 200      |
| DecisionTree                  | 0.570    | 44 359 340       | -0.327      | 231 351 400      |
| DecisionTree (GridSearchCV)  | 0.524    | 49 052 000       | -0.374      | 239 505 500      |
| LinearRegression              | 0.144    | 88 223 000       | 0.006       | 173 284 300      |




### 🏠 Résultats des modèles - Maisons 

| Modèle                         | R² Lille | MSE Lille       | R² Bordeaux | MSE Bordeaux     |
|-------------------------------|----------|------------------|-------------|------------------|
| RandomForest (GridSearchCV)  | 0.064    | 957 047          | -2.956      | 8 187 240        |
| LinearRegression              | 0.031    | 990 380          | -3.041      | 8 361 494        |
| DecisionTree (GridSearchCV)  | 0.017    | 1 004 800        | -0.374      | 8 130 029        |
| RandomForest                  | -0.010   | 1 032 614        | -2.973      | 8 222 158        |
| XGBoost                       | -0.127   | 1 152 406        | -3.164      | 8 616 030        |
| DecisionTree                  | -0.692   | 1 729 826        | -3.518      | 9 349 422        |


### 📌 Analyse 

#### 🏢 Appartements

- Tous les modèles chutent drastiquement en performance sur Bordeaux (R² proches de 0 ou négatifs).
- Le meilleur R² sur Lille est obtenu par XGBoost (0.68), mais il devient négatif sur Bordeaux (-0.17).
- LinearRegression est très faible à l'entraînement, mais a le meilleur score sur Bordeaux (R² ≈ 0.006) — bien que cela reste mauvais.

#### 🏠 Maisons

- Les modèles ont déjà de très faibles R² sur Lille (max ≈ 0.06).
- Sur Bordeaux, aucun modèle ne dépasse -0.37. Le DecisionTree (GridSearchCV) s'en sort "moins
mal" que les autres.
- Tous les modèles montrent une généralisation catastrophique pour les maisons à Bordeaux.

### 🧩 Hypothèses explicatives

Les modèles entraînés sur Lille ne sont pas généralisables à Bordeaux. Cela peut être dû à :

- des caractéristiques socio-économiques ou géographiques très différentes,
- des variables importantes manquantes ou mal encodées (ex : localisation fine, quartier, année
de construction, etc.),
- ou encore à un déséquilibre dans la distribution des prix ou surfaces entre les deux villes.
